File: java/src/main/java/com/logicalclocks/hsfs/metadata/ExpectationsApi.java
/*
 * Copyright (c) 2020 Logical Clocks AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

package com.logicalclocks.hsfs.metadata;

import com.damnhandy.uri.template.UriTemplate;
import com.logicalclocks.hsfs.EntityEndpointType;
import com.logicalclocks.hsfs.FeatureGroup;
import com.logicalclocks.hsfs.FeatureStoreException;
import com.logicalclocks.hsfs.engine.DataValidationEngine;
import lombok.NonNull;
import org.apache.http.HttpHeaders;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPut;
import org.apache.http.entity.StringEntity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import static com.logicalclocks.hsfs.metadata.HopsworksClient.PROJECT_PATH;
import static com.logicalclocks.hsfs.metadata.HopsworksClient.getInstance;

public class ExpectationsApi {

  public static final String ENTITY_ROOT_PATH = "{/entityType}";
  public static final String ENTITY_ID_PATH = ENTITY_ROOT_PATH + "{/entityId}";
  public static final String EXPECTATIONS_PATH =
      ENTITY_ID_PATH + "/expectations{/name}{/predicate}{/feature}{?engine,filter_by,sort_by,offset,limit,expand}";

  private static final Logger LOGGER = LoggerFactory.getLogger(ExpectationsApi.class);

  private EntityEndpointType entityType;

  public ExpectationsApi(@NonNull EntityEndpointType entityType) {
    this.entityType = entityType;
  }

  public Expectation put(FeatureGroup featureGroup, Expectation expectation) throws FeatureStoreException, IOException {
    return put(featureGroup.getFeatureStore().getProjectId(), featureGroup.getFeatureStore().getId(),
      featureGroup.getId(), expectation);
  }

  private Expectation put(Integer projectId, Integer featurestoreId, Integer entityId, Expectation expectation)
      throws FeatureStoreException, IOException {
    HopsworksClient hopsworksClient = getInstance();
    String pathTemplate = PROJECT_PATH + FeatureStoreApi.FEATURE_STORE_PATH + EXPECTATIONS_PATH;

    String uri = UriTemplate.fromTemplate(pathTemplate)
        .set("projectId", projectId)
        .set("fsId", featurestoreId)
        .set("entityType", entityType.getValue())
        .set("entityId", entityId)
        .set("name", expectation.getRule().getName())
        .set("predicate", expectation.getRule().getPredicate())
        .set("feature", expectation.getFeature())
        .expand();

    String expectationStr = hopsworksClient.getObjectMapper().writeValueAsString(expectation);
    HttpPut putRequest = new HttpPut(uri);
    putRequest.setHeader(HttpHeaders.CONTENT_TYPE, "application/json");
    putRequest.setEntity(new StringEntity(expectationStr));

    LOGGER.info("Sending metadata request2: " + uri);
    LOGGER.info(expectationStr);

    return hopsworksClient.handleRequest(putRequest, Expectation.class);
  }

  public List<Expectation> get(FeatureGroup featureGroup) throws FeatureStoreException, IOException {
    return get(featureGroup.getFeatureStore().getProjectId(), featureGroup.getFeatureStore().getId(),
      featureGroup.getId(), null, null, null, null);
  }

  public Expectation get(FeatureGroup featureGroup, Rule.Name name, Rule.Predicate predicate,
      String feature) throws FeatureStoreException, IOException {
    List<Expectation> expectations =
        get(featureGroup.getFeatureStore().getProjectId(), featureGroup.getFeatureStore().getId(),
        featureGroup.getId(), name, predicate, feature, null);
    return !expectations.isEmpty() ? expectations.get(0) : null;
  }

  private List<Expectation> get(Integer projectId, Integer featurestoreId, Integer entityId, Rule.Name name,
      Rule.Predicate predicate, String feature, DataValidationEngine.Engine engine)
      throws FeatureStoreException, IOException {
    String pathTemplate = PROJECT_PATH
        + FeatureStoreApi.FEATURE_STORE_PATH
        + EXPECTATIONS_PATH;

    UriTemplate uriTemplate = UriTemplate.fromTemplate(pathTemplate)
        .set("projectId", projectId)
        .set("fsId", featurestoreId)
        .set("entityType", entityType.getValue())
        .set("entityId", entityId)
        .set("expand","rules");

    if (name != null) {
      uriTemplate.set("name", name);
    }
    if (predicate != null) {
      uriTemplate.set("predicate", predicate);
    }
    if (feature != null) {
      uriTemplate.set("feature", feature);
    }

    if (engine != null) {
      uriTemplate.set("engine", engine);
    }

    String uri = uriTemplate.expand();
    LOGGER.info("Sending metadata request: " + uri);
    HttpGet getRequest = new HttpGet(uri);
    HopsworksClient hopsworksClient = getInstance();
    Expectation dto = hopsworksClient.handleRequest(getRequest, Expectation.class);
    List<Expectation> expectations;
    if (dto.getCount() == null) {
      expectations = new ArrayList<>();
      expectations.add(dto);
    } else {
      expectations = dto.getItems();
    }
    LOGGER.info("Received expectations: " + expectations);
    return expectations;
  }

}

File: python/hsfs/core/expectations_api.py
#
#   Copyright 2020 Logical Clocks AB
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#

from hsfs import client, expectation


class ExpectationsApi:
    def __init__(self, feature_store_id, entity_type=None):
        """Expectations endpoint for `featurestores` and `featuregroups` resource.

        :param feature_store_id: id of the respective featurestore
        :type feature_store_id: int
        :param entity_type: "featuregroups"
        :type entity_type: str
        """
        self._feature_store_id = feature_store_id
        self._entity_type = entity_type

    def create(self, expectation):
        """Create and Feature Store expectation or Attach it by name to a Feature Group.

        :param expectation: expectation object to be created for a feature store
        :type expectation: `Expectation`
        """
        _client = client.get_instance()
        path_params = [
            "project",
            _client._project_id,
            "featurestores",
            self._feature_store_id,
            "expectations",
        ]

        headers = {"content-type": "application/json"}
        payload = expectation.json() if expectation else None
        _client._send_request("PUT", path_params, headers=headers, data=payload)

    def attach(self, feature_group, name):
        """Attach a Feature Store expectation to a Feature Group.

        :param feature_group: metadata object of the instance to attach the expectation to
        :type feature_group: FeatureGroup
        :param name: name of the expectation to be attached
        :type name: str
        """
        _client = client.get_instance()
        path_params = [
            "project",
            _client._project_id,
            "featurestores",
            self._feature_store_id,
            self._entity_type,
            feature_group.id,
            "expectations",
            name,
        ]

        _client._send_request("PUT", path_params)

    def delete(self, name):
        """Delete a Feature Store expectation.

        :param name: name of the expectation to be deleted
        :type name: str
        """
        _client = client.get_instance()
        path_params = [
            "project",
            _client._project_id,
            "featurestores",
            self._feature_store_id,
            "expectations",
            name,
        ]

        _client._send_request("DELETE", path_params)

    def detach(self, feature_group, name):
        """Detach a Feature Store expectation from a Feature Group.

        :param feature_group: metadata object of the instance to attach the expectation to
        :type feature_group: FeatureGroup
        :param name: name of the expectation to be attached
        :type name: str
        """
        _client = client.get_instance()
        path_params = [
            "project",
            _client._project_id,
            "featurestores",
            self._feature_store_id,
            self._entity_type,
            feature_group.id,
            "expectations",
            name,
        ]

        _client._send_request("DELETE", path_params)

    def get(self, name=None, feature_group=None):
        """Get the expectations of a feature store or feature group.

        Gets all feature store expectations if no feature group is specified.
        Gets all feature store or feature group expectations if no name is specified.

        :param name: expectation name
        :type name: str
        :param feature_group: feature group to get the expectations of
        :type feature_group: FeatureGroup
        :return: list of expectations
        :rtype: list of dict
        """
        _client = client.get_instance()
        path_params = [
            "project",
            _client._project_id,
            "featurestores",
            self._feature_store_id,
        ]

        if feature_group is not None:
            path_params.extend([self._entity_type, feature_group.id, "expectations"])
        else:
            path_params.append("expectations")

        if name:
            path_params.append(name)

        return expectation.Expectation.from_response_json(
            _client._send_request("GET", path_params)
        )
